### PR TITLE
feat: improved accessibility scanning in CI

### LINKED ISSUE #782: Improve accessibility scan in CI
Closes PR #831 

Include items missing or not complete in PR #654 
- [x] tests
- [ ] bucket in Safe Inputs GCP project (bucket creation yaml linting failed)
- ~~[ ] Swap out pause in Cloudbuild for health check. (The more efficient image for axe-testing finishes before the docker compose environment has completed spinning up.)~~ Stephen fixed in PR #654 
- ~~[ ] Swap filename for SHA when run in cloudbuild~~ incorporated in PR #654 
- ~~[ ] Look at testing for GoC specific requirements i.e. are Aria labels in both official languages present~~
- [x] https://github.com/HC-IPPM/safe-inputs/pull/654#discussion_r1805092872 : 
* "some routes only exist if the right data is present (e.g. links to collection management or upload routes from the home page only exist if the logged in user has collections to manage/upload to)
could be addressed by seeding the database with specially crafted data before performing the accessibility scan. It'll be a small maintenance burden, but it's that or try and make a crawler that can correctly fill out and submit valid forms, and then re-crawl the site for dynamic changes/new forms until everything has been reached... which is going to be slower and a nightmare to build, haha.
* navigation can also occur via button clicks (e.g. when the server response includes a redirect, which we won't generally do, or when client side JS navigates the client based on server response, such as moving the user the the collection management route for a new collection after submitting a collection creation form)
off the top of my head, and ideally it stays this way, any non-link navigation leads to a page that is also reachable by link, if the data for it exists. Might be able to ignore this problem 🤞
* elements that render/respond to user input won't be interacted with by a shallow scan on each route (e.g. owner and uploader email input fields don't exist until the user clicks the button to add new emails)
this is also likely not worth trying to fully automate as part of our end-to-end scans, that's too hard a problem. Likewise, maintaining puppeteer scripts to exercise dynamic elements on each page would be a big maintenance burden. I'd prefer to leave interactive elements out of scope in this end-to-end crawler, and instead figure out a unit test (component level) equivalent for more dynamic accessibility checks."

First bullet to be prioritized.

### PR DESCRIPTION
**Pull request by LilaKelland**
closes #782 

- [x] Add tests
- [x] Seed database to activate all routes 
- ~~[ ] Add test for bilingual aria labels~~ (moved to PR #865)
- [x] Add paragraph to devsecops readme (unless we'd like to include more for each....)
- [x]  refactor
- [x] move axe-testing into the devsecops folder 

Couple of things to note:
*  The results will be saved into the safe-inputs-devsecops-outputs-for-dashboard which currently has a lifecycle of 30 days. 
* If this is an okay set up - will need to add the devsecops cloudbuild trigger 

Sorry lots of changes!  Let me know if you have any questions.

### CODE REVIEW

File: api/src/seed_dev_data.ts
Comment (LilaKelland) [inline api/src/seed_dev_data.ts:1]: Had copied 'insert records' from your code above, curious why its done in two steps, I tried combining them and it didn't end up triggering the scan of 'upload-records'.

File: axe-testing/package-lock.json
Comment (LilaKelland) [inline axe-testing/package-lock.json:1]: These files were moved into devsecops/axe-testing

File: devsecops/axe-testing/Dockerfile.tests
Comment (LilaKelland) [inline devsecops/axe-testing/Dockerfile.tests:1]: This is just used for tests (runs both unit and the one e2e), but it's basically the same as the Dockerfile for the application, with the exception of the CMD.  I'm using a dockerfile so there's no need to install puppeteer locally, but it seems like duplication - is there a better way to do this?

File: api/src/seed_dev_data.ts
Comment (Stephen-ONeil) [reply api/src/seed_dev_data.ts:1]: Combining which steps and how exactly?

File: devsecops/axe-testing/Dockerfile.tests
Comment (Stephen-ONeil) [reply devsecops/axe-testing/Dockerfile.tests:1]: If only the `CMD` differs, you can pass an override command option to `docker run`. Check `docker run --help`, the command, note it's not as an options flag, you just include the command following the image name.

File: api/src/seed_dev_data.ts
Comment (LilaKelland) [reply api/src/seed_dev_data.ts:1]: Sorry that wasn't clear at all.   There's now 2 insert_records (that works and populates the data to trigger the scan on the manage_records pages) - this is the same as user_1/ owner_1 above in this file)

 await insert_records(user_axe, collection_axe, [
  { column_1: 1, column_2: 2 },
  { column_1: 3, column_2: 4 },
  { column_1: 5, column_2: 6 },
]);

await insert_records(user_axe, collection_axe, [
  { column_1: 7, column_2: 8 },
  { column_1: 9, column_2: 10 },
  { column_1: 11, column_2: 12 },
]); 

 but when I tried to combine them into one insert_records:  

await insert_records(user_axe, collection_axe, [
  { column_1: 1, column_2: 2 },
  { column_1: 3, column_2: 4 },
  { column_1: 5, column_2: 6 },
  { column_1: 7, column_2: 8 },
  { column_1: 9, column_2: 10 },
  { column_1: 11, column_2: 12 },
]);

It didn't trigger the scan of the additional pages.  I'm assuming I may have had a syntax error or something like that, but was wondering if there was different specific reason it was spilt out above with the owner_1/user_1?

File: devsecops/axe-testing/Dockerfile.tests
Comment (LilaKelland) [reply devsecops/axe-testing/Dockerfile.tests:1]: Very cool!  On second glance, the testing one isn't running as root and installs the extra test files and config, so slightly different..... maybe I'll try to run tests as non-root tomorrow and see about combining them - thanks!

File: api/src/seed_dev_data.ts
Comment (Stephen-ONeil) [reply api/src/seed_dev_data.ts:1]: I believe I did two separate inserts to have some variation in the creation time stamps, shouldn't have any other impact, hmm.

There also aren't currently any routes that are dependent on the existence of columns either. The `upload-records` route placeholder should exist as long as any collections exist for the signed in user (either as owner or uploader). 

Syntax looks fine too. Could you try to recreate it and let me know if you still have the same issue? Could be something weird that's not occurring to me, would be good to know about, thanks!

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
> If this is an okay set up - will need to add the devsecops cloudbuild trigger
I think it can be part of the root cloud build trigger :+1: 

The e2e test is more of a "dependency test" right now, just asserting that puppeteer and axe work together. Dependency tests can be worth doing in some cases, but it'd be ideal here for the end to end test to exercise crawl-page.js and process-axe-report.js altogether, and then assert that the final output is as expected. It'll test the dependencies in the process, but also that the real output of a `AxePuppeteer` run works with your other steps. :+1: 

Parts of `index.js` might need to be recreated, without the safe-inputs specific portions, to do this, right? That'd be fine for now, but you can take challenges like that when writing tests as an indication that the structure of the code itself can be refine,  i.e. in this case, to separate run time context (safe inputs) specific code/configuration from the abstract function of the axe-testing code (crawling a generic site and noting a11y issues).

 Happy to discuss further on a call!

File: devsecops/axe-testing/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "dbg": "node --experimental-vm-modules --inspect-brk node_modules/.bin/jest --runInBand --no-cache",
+    "prestart:docker": "mkdir -p axe-results && sudo chmod 777 axe-results && docker build -t axe .",
+    "start": "node --experimental-vm-modules index.js",
+    "start:docker": "docker run --rm --network host -v ${PWD}/axe-results:/app/axe-results -e COMMIT_SHA=$(git rev-parse HEAD) axe",
+    "test:unit": "NODE_OPTIONS=--experimental-vm-modules jest ./src/*.test.js --coverage",
+    "test:e2e:startServer": "npx http-server ./tests/test-pages -p 8080 & echo $! > server.pid",
+    "test:e2e:waitForSeverToStart": "for i in {1..10}; do nc -z localhost 8080 && break || echo 'Waiting for server to be ready...'; sleep 1; done",
+    "test:e2e:runTests": "NODE_OPTIONS=--experimental-vm-modules npx jest tests/e2e.test.js",
+    "test:e2e:stop": "kill $(cat server.pid) || true",
+    "test:e2e": "npm run test:e2e:startServer && npm run test:e2e:waitForSeverToStart && npm run test:e2e:runTests && npm run test:e2e:stop",
```
Comment (Stephen-ONeil) [inline devsecops/axe-testing/package.json:18]: Should be able to start the http server as a pre-test step inside the e2e test, and clean it up at the end, I think!

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
Sorry, vacation last week wiped my brain, I know we discussed possibly reorganizing the axe testing entry point to isolate the safe-inputs details (login flow) from the generic crawling and axe scanning code. Was there anything else that we discussed as possibly in scope for this PR?

Looking at it with a "fresh" mind (for better or worse :sweat_smile:) I'm fine with leaving that for a further follow-up step, perhaps closer to when we think this axe scanning pattern might mature in to something standalone/reusable. Only one thing that jumped right out to me otherwise.

File: devsecops/axe-testing/jest.config.cjs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  transform: {
+    '.*\\.js$': 'babel-jest',
+  },
+  testTimeout: 60000, // Increase timeout for e2e test
```
[the remark this answers is not in the document]
Comment (LilaKelland) [reply devsecops/axe-testing/jest.config.cjs:5]: Thanks!  I'll take a look at this tomorrow.  I think it was mostly the entry point, but that sounds good.

**Review by Stephen-ONeil — APPROVED** [review]
Last optional thought, feel free to take it or not but approved either way. Now that there's separate e2e and unit test config files, you could encode test type in the file paths and match on that, e.g. `*.test.e2e.js` and `*.test.unit.js`. 

It might also be worth renaming `devsecops/axe-testing/tests` to `.../e2e-tests` too, just so future devs don't get confused about what tests live where.

**Comment by LilaKelland** [comment]
Thanks Stephen!  That's a good idea - changed the names but then my modifications to eslint.config.mjs for the new test file names weren't working, so switched back to .eslintrc (and sending back for review before merging)

**Pull request merged**